source/jsonParsing.ml

```diff
@@ -64,3 +64,201 @@ let optional_list_member ~f name json =
 let string_list_member = list_member ~f:to_string
 
 let path_list_member = list_member ~f:to_path
+
+module Location = struct
+  type t = {
+    start_line: int;
+    end_line: int;
+    start_column: int;
+    end_column: int;
+  }
+  [@@deriving show, eq]
+
+  let from_decoded_range range =
+    let (start_line, end_line), (start_column, end_column) = range in
+    { start_line; end_line; start_column; end_column }
+end
+
+module Node = struct
+  type 'a t = {
+    location: Location.t option;
+    value: 'a;
+  }
+  [@@deriving show, eq]
+end
+
+exception
+  JsonmException of {
```

Comment:
Is this only used on parse errors? Could we use a result type instead of exceptions?
We can still have the exception internally to implement parsing, but I would prefer result types in the top level API.

Reply:
Yes only only parse errors. I made json_of_string return result type now - hoping that's what you meant :)

Reply:
> I tried to return an error of type TaintConfiguration.Error.t but then cyclic dependency problem.

What is the cyclic dependency? TaintConfiguration.Error.t should depend on JsonAst, that's fine. but I don't see why JsonAst could depend on TaintConfiguration.Error.t.

Reply:
I tried to open `TaintConfiguration.Error` on `JsonAst` to return `TaintConfiguration.Error` directly in `JsonAst`

Reply:
Resolved this by introducing a new Error module. I think this makes sense as we go forward and need a error type for some utility functions?

Reply:
> I tried to open TaintConfiguration.Error on JsonAst to return TaintConfiguration.Error directly in JsonAst

Well, that's the problem. `JsonAst` should be an independent library that does not depend on `TaintConfiguration.Error`.

> Resolved this by introducing a new Error module. I think this makes sense as we go forward and need a error type for some utility functions?

Yes that sounds like the right approach.